File: Backend/asp.netcore/Modules/WebSVC.cs
using Microsoft.AspNetCore.Http;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Web.Application.Interfaces;
using Web.Application.Lib;
using Web.Application.Services;
using Web.Application.Services.DB;
using Web.Appliction.Ext;

namespace Web.Application.Modules
{
    public class WebSVC : IModule
    {
        public string Authenticated(HttpContext context)
        {
            return JsonConvert.SerializeObject(
                new { bearer = context.Response.Headers["Authorization"] }
            );
        }

        public string Process(HttpContext context)
        {
            string result = null;

            // get relative path
            var navigation = (IDictionary<string, object>)context.Items["navigation"];
            string navigationPath = navigation.Get("url")?.ToString();

            // db
            var db = (SQL)context.Items["db"];

            var regex = new Regex(Regex.Escape(navigationPath));
            string[] relativePaths = regex
                .Replace(context.Request.Path.ToString(), "", 1)
                .Split("/");

            // get websvc name            
            string websvcName = string.Empty;
            if (relativePaths.Length > 1)
                websvcName = relativePaths[1];

            // get navigation id
            string navigationId = navigation.Get("_id")?.ToString();

            // load web services
            if (
                string.IsNullOrEmpty(websvcName) == false
                && string.IsNullOrEmpty(navigationId) == false
                )
            {
                var serviceParam = new Dictionary<string, object>();
                serviceParam["navigation_id"] = navigationId;
                serviceParam["api_url"] = websvcName;
                var service = db.Query(
                    @"SELECT * FROM core_websvc WHERE navigation_id=@navigation_id AND api_url=@api_url"
                    , serviceParam
                ).First();
                
                // services exists and provides api for the method
                string method = context.Request.Method.ToLower();
                if (service != null)
                {
                    // load workflow
                    string workflow_id = service.Get($"{method}_workflow")?.ToString();
                    if (string.IsNullOrEmpty(workflow_id) == false)
                    {
                        var workflowParam = new Dictionary<string, object>();
                        workflowParam["_id"] = workflow_id;
                        var workflow = db.Query(
                            @"SELECT * FROM core_workflow WHERE _id=@_id"
                            , workflowParam
                        ).First();
                        if (workflow != null)
                        {
                            // Prepare inputs to the script
                            IDictionary<string, object> Inputs = new Dictionary<string, object>();

                            // load datasources
                            var dataservices = new List<object>();
                            var dsIds = service.GetArray($"{method}_datasource") as object[];
                            foreach (object dsId in dsIds)
                            {
                                if (string.IsNullOrEmpty($"{dsId}") == false)
                                {
                                    // load datasource
                                    var dataserviceParam = new Dictionary<string, object>();
                                    dataserviceParam["_id"] = dsId;
                                    var dataservice = db.Query(
                                        "SELECT * FROM core_dataservice WHERE _id=@_id"
                                        , dataserviceParam).First();

                                    var createdDS = new SQL($"{dataservice.Get("connectionString")}");

                                    // instantiate                             
                                    dataservices.Add(createdDS);
                                }
                            }

                            // produce entire script                                                         
                            object scriptResult = Script.Run(
                                context
                                , $"{workflow?.Get("_id")}"
                                , $"{service.Get($"{method}_configuration")}"
                                , dataservices);

                            // if script result is string
                            if (scriptResult != null && scriptResult.GetType() == typeof(string))
                            {
                                result = scriptResult.ToString();
                                // jsonp                                
                                string callback = WebTools.Get(context, "callback");
                                if (string.IsNullOrEmpty(callback) == false)
                                    result = $"{callback}({result})";
                            }

                            // if script result is anonymous type
                            else if (scriptResult != null && scriptResult.GetType().Name.Contains("AnonymousType"))
                            {
                                // return json serialized format
                                result = JsonConvert.SerializeObject(scriptResult);
                            }

                            // if script result is not string
                            else if (scriptResult != null)
                            {
                                // response with token
                                byte[] byteResult = (byte[])scriptResult;
                                context.Response.Body.WriteAsync(byteResult, 0, byteResult.Length);
                            }

                        }
                    }

                }
            }

            return result;
        }
    }
}

File: Backend/php/modules/websvc.php
<?php
if(!defined('prevent_direct_access')) { header('HTTP/1.0 403 Forbidden'); exit; }

class websvc {
    public function RequiresAuthentication($context) {
        return true;
    }

    public function Process($context) {        
        $result = null;

        // get web service name
        $regex = '/'.preg_quote($context->nav['url'], '/').'/';
        $relativePaths = preg_replace($regex, '', strtok($_SERVER["REQUEST_URI"],'?'), 1);
        $websvcName = str_replace('/', '', $relativePaths);
        
        if(IsNullOrEmpty($websvcName) == false) {
            // load web services                        
            $sql = "SELECT * FROM core_websvc WHERE navigation_id=? AND api_url=?";            
            $result = $context->conn->query($sql, array($context->nav['_id'], $websvcName));
            if(count($result) == 0) {
                header('HTTP/1.0 403 Forbidden');                
                return 'web service not found '.$context->nav['_id'].' '.$websvcName;
            }
                
            if (count($result) > 0) {
                $websvc = $result[0];
                $method = strtolower($_SERVER['REQUEST_METHOD']);
                if($websvc) {
                    // load workflow
                    $workflow_id = $websvc[$method.'_workflow'];
                    if(IsNullOrEmpty($workflow_id) == false) {
                        $sql = "SELECT * FROM core_workflow WHERE _id=?";
                        $result = $context->conn->query($sql, array($workflow_id));
                        if (count($result) > 0) {
                            $workflow = $result[0];

                            // configurtion
                            $config = $websvc[$method.'_configuration'];                        

                            // load data source
                            $ds = null;
                            $ds_id = $websvc[$method.'_datasource'];
                            if(IsNullOrEmpty($ds_id) == false) {
                                $sql = "SELECT * FROM core_dataservice WHERE _id=?";
                                $result = $context->conn->query($sql, array($ds_id));
                                if (count($result) > 0) {
                                    $dataservice = $result[0];
                                    $ds = new mysql($dataservice['connectionString']);
                                }
                            }
                            
                            // run script
                            $script = $workflow['script'];                          
                            $script = str_replace('<?php', '', $script);
                            $script = str_replace('?>', '', $script);
                            $result = eval($script);                            
                        }
                    }
                }
            }
        }                        

        return $result;
    }

    public function Authenticated($context) {
        
    }
}

?>